### ISSUE TITLE
Is it possible to add fonts?

### ISSUE THREAD

**Issue by bbigras** [t184256/nix-on-droid#120]
I'm using `fonts.fontconfig.enable = true;` and:

```nix
{
  home.packages = with pkgs; [
    meslo-lgs-nf
  ];
}
```

and I seem to be missing fonts with the powerlevel10k prompt.

**Comment by t184256** [comment]
TL;DR: you can set one font

Fonts are hard to do right. I've started twice or thrice, never finished.

Font facts:

* in normal nix-on-droid usage (just running the terminal), none of the linux/home-manager font configuration facilities matter, fontconfig is ignored
* the only thing that renders fonts is [my fork of the Termux app](https://github.com/t184256/nix-on-droid-app)
* the fork has the styling addon baked in, so it looks for a `~/.termux/font.ttf`
* so, good news are, you can change the font
* bad news are, it has to be a regular file and not a symlink to `/nix/store` because, as far as the terminal app is concerned, there's no `/nix/store`. Auto-copying it on activation sounds hacky. Patching the app to do the sort of broken symlink resolution we already do for login script is definitely an option, but I'm basically allergic to Android development. PRs are welcome though.
* another bad thing is, Termux terminal emulator isn't exactly breathtaking with its font-rendering capabilities, even doing obliques instead of italics (my poor font)
* finally, there's no family detection, fallback chain or anything, none of this stuff. one ttf to rule them all

So, if you want a font with extra symbols, you'd have to find a TTF that has all the symbols you want in one file and then put it at `~/.termux/font.ttf`.

**Comment by bbigras** [comment]
![Screenshot_20210624-121308.png](https://user-images.githubusercontent.com/24027/123297518-2a886000-d507-11eb-9bc1-ff2a267206d3.png)

It works! Thanks!

**Comment by PaulGrandperrin** [comment]
@t184256 FYI, the people of  `nix-darwin` are doing this "hacky auto-copying on activation": https://github.com/LnL7/nix-darwin/blob/master/modules/fonts/default.nix#L55

I think the ease of use is so much worth this "hackyness".

Here's my code for those who want (using `home-manager`'s lib):
```nix
home.activation = {
  copyFont = lib.hm.dag.entryAfter ["writeBoundary"] ''
    $DRY_RUN_CMD install $VERBOSE_ARG -D "${pkgs.nerdfonts.override { fonts = [ "FiraCode" ]; }}/share/fonts/truetype/NerdFonts/Fira Code Regular Nerd Font Complete Mono.ttf" ${config.home.homeDirectory}/.termux/font.ttf
  '';
};
```

**Comment by PaulGrandperrin** [comment]
Similar code in `home-manager` for the same issue: https://github.com/nix-community/home-manager/blob/781d25b315def05cd7ede3765226c54216f0b1fe/modules/targets/darwin/fonts.nix#L14=

**Comment by t184256** [comment]
I don't like copying each time over and over on each activation (and I also don't like modding the apk to resolve a symlink that doesn't make sense outside of the proot environment). Any more lightweight options?

**Comment by PaulGrandperrin** [comment]
I don't like that too, my `nerdfont/firacode` is quite heavy at 1M.
Instead of copying, I would do like what the `nix-darwin` devs are doing: using `rsync` https://github.com/LnL7/nix-darwin/blob/master/modules/fonts/default.nix#L56=
This way, the file should be copied only if it has changed.
However, I think they forgot to use the `--checksum` flag because the `rsync` heuristics are using the modification time (and size) and in the nix store, it is always at the unix epoch. so if a font is updated by but it still the same size, rsync will miss it.

More info: https://serverfault.com/questions/211005/rsync-difference-between-checksum-and-ignore-times-options

TLDR: I think using `rsync --checksum` solve this problem quite nicely. Will only incur 2 times the font size sequential read at each activation. Which is cheap on flash and doesn't cause wear-leveling.

**Comment by t184256** [comment]
rsync is a heavy dependency for basically just checksumming or comparing two files

**Comment by PaulGrandperrin** [comment]
`rsync` is 645K on arm64, but yes, it can also be done with sha1sum and cp

**Comment by PaulGrandperrin** [comment]
there's also `cmp` from `diffutils` which is more appropriate to compare files than doing 2 checksums, but it weights 1.9M, so I guess it's a no-go..

**Comment by PaulGrandperrin** [comment]
One thing to note is that this dependency would only be pulled is the user choose to install a font, which will probably weight more than `rsync` itself.

**Comment by PaulGrandperrin** [comment]
Version using `sha1sum`s and `install`:
```nix
home.activation = {
  copyFont = let
      font_src = "${pkgs.nerdfonts.override { fonts = [ "FiraCode" ]; }}/share/fonts/truetype/NerdFonts/Fira Code Regular Nerd Font Complete Mono.ttf";
      font_dst = "${config.home.homeDirectory}/.termux/font.ttf";
    in lib.hm.dag.entryAfter ["writeBoundary"] ''
      test $(sha1sum "${font_src}"|cut -d' ' -f1) != $(sha1sum "${font_dst}" |cut -d' ' -f1) && $DRY_RUN_CMD install $VERBOSE_ARG -D "${font_src}" "${font_dst}"
  '';
};
```
